### PR TITLE
feat: use `objcopy` instead of `LIEF`

### PR DESCRIPTION
**Pull request by RaisinTen**
Fixes: https://github.com/nodejs/postject/issues/63
Signed-off-by: Darshan Sen <raisinten@gmail.com>

### CODE REVIEW

File: test/cli.mjs [hunk 1]
```diff
@@ -95,7 +112,6 @@ describe("postject CLI", () => {
           console.log(err.message);
         }
         if (codesignFound) {
-          execSync(`codesign --sign - ${filename}`);
```
Comment (RaisinTen) [inline test/cli.mjs:98]: For some reason, code signing the binary after running `objcopy --add-section` on macOS causes the binary to crash:
```sh
$ ./node
[1]    5366 killed     ./node
$ echo $?
137
```

and I can't even `lldb` into it:

```sh
lldb -- ./node
(lldb) target create "./node"
Current executable set to '/Users/raisinten/Desktop/git/postject/node' (x86_64).
(lldb) run
error: Cannot allocate memory
(lldb) ^D
```

However, it seems to be possible to run `objcopy` after code signing and not invalidate the code signature! I also tried to check if the new section was being added after the `LC_CODE_SIGNATURE` section that @bnoordhuis mentioned in https://github.com/nodejs/node/pull/45066#issuecomment-1287431258 and it seems like `objcopy` adds new sections at the end, even after `LC_CODE_SIGNATURE` but the `codesign` tool thinks that this is valid:

```txt
$ otool -l ./node | tail -n29
Load command 18
      cmd LC_CODE_SIGNATURE
  cmdsize 16
  dataoff 84875008
 datasize 663232
Load command 19
      cmd LC_SEGMENT_64
  cmdsize 152
  segname __POSTJECT
   vmaddr 0x00000001057d4000
   vmsize 0x0000000000001000
  fileoff 68509696
 filesize 4096
  maxprot 0x00000007
 initprot 0x00000007
   nsects 1
    flags 0x0
Section
  sectname __NODE_JS_CODE
   segname __POSTJECT
      addr 0x00000001057d4000
      size 0x0000000000000014
    offset 68509696
     align 2^0 (1)
    reloff 0
    nreloc 0
     flags 0x00000000
 reserved1 0
 reserved2 0
```

```sh
$ codesign -v ./node
$ echo $?
0
```

Any clue what's happening here?

File: src/api.js [hunk 2]
```diff
@@ -39,30 +43,10 @@ async function inject(filename, resourceName, resourceData, options) {
...
-        if (result === postject.InjectResult.kAlreadyExists) {
-          throw new Error(
-            `Segment and section with that name already exists: ${machoSegmentName}/${sectionName}\n` +
-              "Use --overwrite to overwrite the existing content"
-          );
-        }
-      }
-      break;
+      execSync(
+        `/usr/local/opt/llvm/bin/llvm-objcopy --add-section ${machoSegmentName},__${resourceName}=${resource} ${filename}`
```
Comment (robertgzr) [inline src/api.js:47]: can we just use `llvm-objcopy` from `$PATH`?

Comment (RaisinTen) [reply src/api.js:47]: `brew install llvm` does not add it to the path

Comment (robertgzr) [reply src/api.js:47]: hmm. this will probably not work on linux, my system the llvm install is located under `/usr/lib/llvm/14/`

**Comment by jviotti** [comment]
@RaisinTen I thought the agreement from the last call is that we would NOT do this on Postject.

**Comment by RaisinTen** [comment]
@jviotti we didn't consider a point during the meeting. I posted about it yesterday in the issue - https://github.com/nodejs/postject/issues/63#issuecomment-1300247266.

File: src/api.js (hunk 2 above)
Comment (RaisinTen) [reply src/api.js:47]: correct, the path (even the binary name) is different on linux. I haven't implemented that part yet, will do that soon.

[the remark this answers is not in the document]
Comment (RaisinTen) [reply src/api.js]: good catch, done!

**Comment by jviotti** [comment]
I thought the conclusion was:

- We want Postject to inject across platforms (you raised valid use cases like M1 not being available on CI, etc)
- The first point is a must for integrating Postject to i.e. PKG
- My though was to consider the LLVM approach only for testing purposes within Node.js core as an alternative to integrating Postject under `test/fixtures`

cc @tony-go @mhdawson in case you can confirm whether I got it wrong or not

**Comment by RaisinTen** [comment]
Correct but I forgot to mention a point during the meeting - https://github.com/nodejs/postject/issues/63#issuecomment-1300247266:
> the cross-platform binary creation might actually not even work if your project has native bindings

i.e., while testing out pkg I found that if it produces an ELF file when you're packaging an app with native dependencies on macOS, the `.node` files would still be Mach-O binaries, so it won't run on Linux

Does this change anything?

**Comment by jviotti** [comment]
Yeah, that's very interesting. If so, I'm fine with the `objcopy` approach, as users will need to build the CLI on each platform anyway to cover all possible edge cases (i.e. native add-ons).

**Review by dsanders11 — CHANGES_REQUESTED** [review]
Left some comments about the implementation on this PR, but I’m generally -1 on this change, see my comments on #63.

Another general comment on this PR, the commit message isn’t accurate for what the PR does - first it should specifically mention `llvm-objcopy`, and second it should specify that this PR is macOS only.

**Pull request closed without merging**